File: src/main/java/itv.momo/ProductContainer.java
package itv.momo;

import java.util.HashMap;
import java.util.Map;

interface IProductContainer {
    public int add(Item item) throws Exception;
    public void add(int idx, Item item) throws Exception;
    public void update(int idx, Item item) throws Exception;
    public Item getItem(int idx) throws Exception;
    public Item[] getAllItem();
    public int estimate(HashMap<Integer, Integer> itemCnt);
}

public class ProductContainer implements IProductContainer {
    private final Item[] data;
    private int cnt;


    public ProductContainer(int size) {
        this.data = new Item[size];
        this.cnt = 0;
    }

    public ProductContainer() {
        this.data = new Item[4];
        this.cnt = 0;
    }

    public int add(Item item) throws Exception {
        if (cnt == data.length) {
            throw new Exception(ExceptionMsg.FullItem);
        }

        for (int i = 0; i < data.length; i++) {
            if (data[i] == null) {
                _add(i, item);
                return i;
            }
        }

        return -1;
    }

    public void add(int idx, Item item) throws Exception {
        if (item == null) {
            throw new Exception(ExceptionMsg.AddInvalidItem);
        }

        if (data[idx] != null) {
            throw new Exception(ExceptionMsg.AddItemInExitedIndex(idx));
        }

        _add(idx, item);
    }

    public void update(int idx, Item item) throws Exception {
        if (item == null) {
            throw new Exception(ExceptionMsg.UpdateInvalidItem);
        }

        if (0 < idx || idx >= data.length) {
            throw new Exception(ExceptionMsg.UpdateInvalidIndex(idx));
        }

        if (data[idx] != null) {
            remove(idx);
        }
        _add(idx, item);
    }


    protected void _add(int idx, Item item) {
        cnt += 1;
        data[idx] = item;
    }

    protected Item remove(int idx) {
        cnt -= 1;
        Item item = data[idx];
        data[idx] = null;
        return item;
    }

    public Item getItem(int idx) throws Exception {
        if (idx >= data.length) {
            throw new Exception(ExceptionMsg.SelectNonExitItem);
        }
        return data[idx];
    }

    @Override
    public Item[] getAllItem() {
        return data;
    }

    @Override
    public int estimate(HashMap<Integer, Integer> itemCnt) {
        int sum = 0;
        for (Map.Entry<Integer, Integer> entry : itemCnt.entrySet()) {
            Item item = null;
            try {
                item = getItem(entry.getKey());
            } catch (Exception ignored) {

            }
            if (item != null)
                sum += item.getPrice() * entry.getValue();
        }
        return sum;
    }
}

File: src/main/scala/itv2/ProductContainer.scala
package itv2

import itv.momo.ExceptionMsg

trait IProductContainer {
    def addProduct(item: Item): Int
    def getItem(idx: Int): Item
    def getAllProduct(): Array[Item]
    def estimate(itemCnt : Map[Int, Int]): Int
}

class ProductContainer(size: Int = 4) extends IProductContainer {
    private val data = Array.fill[Item](size)(null)
    private var count = 0;

    override def addProduct(item: Item) = {
        if (count == size) {
            throw new Exception(ExceptionMsg.FullItem)
        }
        val idx = data.indexOf(null)
        data(idx) = item
        idx
    }

    override def getItem(idx: Int) = {
        if (idx < 0 || idx >= size) {
            throw new Exception(ExceptionMsg.SelectNonExitItem)
        }
        data(idx)
    }

    override def getAllProduct(): Array[Item] = data

    override def estimate(itemCnt: Map[Int, Int]) = {
        var sum = 0
        for ((idx, cnt) <- itemCnt) {
            sum += cnt * data(idx).price
        }
        sum
    }
}
